### ISSUE TITLE
Simple slice fusion fails to segment

### ISSUE THREAD

**Issue by jacobhinkle** [NVIDIA/Fuser#1541]
Minimal repro:
```python
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1], contiguity=[True], dtype=DataType.BFloat16)
    T1 = fd.ops.neg(T0)
    T2 = fd.ops.slice(T1, start_indices=[0], end_indices=[5])
    fd.add_output(T2)

with FusionDefinition() as fd:
    nvfuser_fusion_id1(fd)

inputs = [
    torch.randn((10,), dtype=torch.bfloat16, device='cuda:0')
]
fd.execute(inputs)
```
This fails with `Can not find a scheduler to schedule fusion segment`. `NVFUSER_DUMP=segmenter_logging` shows:
```
***Runtime***: Try to schedule fusion un-segmented:

Scheduler _no_op_ ***rejected*** because : output has a concrete dimension
Scheduler _reduction_ ***rejected*** because : No reduction op to schedule
Scheduler _transpose_ ***rejected*** because : First input of SliceOp must be fusion input.
Scheduler _pointwise_ ***rejected*** because : First input of SliceOp must be fusion input.
Scheduler _inner_persistent_ ***rejected*** because : needs a reduction op
Scheduler _outer_persistent_ ***rejected*** because : needs a reduction op
Scheduler _inner_outer_persistent_ ***rejected*** because : needs a reduction op
Scheduler _matmul_ ***rejected*** because : Matmul scheduler supports fusions only with a single MMA op, got: 0

**Segmenter** Considering fusion:
i13 = fmin(i0, 5);
i15 = fmax(0, i13);
T2_g[ iS4{( i0 + ( ( fmax(0, ( fmin(i0, 5) )) ) + ( -i0 ) ) )}rf ]
   = slice( T1_g[ iS1{i0} ], { {0, i15, 1} } )

Scheduler _no_op_ ***rejected*** because : output has a concrete dimension
Scheduler _reduction_ ***rejected*** because : No reduction op to schedule
Scheduler _transpose_ ***rejected*** because : cannot find two mismatching inner most dimensions
Scheduler _transpose_ ***rejected*** because : cannot find two mismatching inner most dimensions
***Accepted*** as: pointwise

**Segmenter** Considering fusion:
T1_g[ iS1{i0} ]
   = -T0_g[ iS0{i0} ];
i13 = fmin(i0, 5);
i15 = fmax(0, i13);
T2_g[ iS4{( i0 + ( ( fmax(0, ( fmin(i0, 5) )) ) + ( -i0 ) ) )}rf ]
   = slice( T1_g[ iS1{i0} ], { {0, i15, 1} } )

Scheduler _no_op_ ***rejected*** because : output has a concrete dimension
Scheduler _reduction_ ***rejected*** because : No reduction op to schedule
Scheduler _transpose_ ***rejected*** because : First input of SliceOp must be fusion input.
Scheduler _pointwise_ ***rejected*** because : First input of SliceOp must be fusion input.
Scheduler _inner_persistent_ ***rejected*** because : needs a reduction op
Scheduler _outer_persistent_ ***rejected*** because : needs a reduction op
Scheduler _inner_outer_persistent_ ***rejected*** because : needs a reduction op
Scheduler _matmul_ ***rejected*** because : Matmul scheduler supports fusions only with a single MMA op, got: 0
An error occurred while executing nvFuser FusionDefinition 1.
```

**Comment by naoyam** [comment]
This seems like because a no-op segment becoming a non no-op segment due to the special handling of uop-only input exprs. A segment with just the slice op is detected as a no-op segment, and then when the whole fusion is finalized, the neg op is prepended to the segment, which means the segment is no longer no-op.

While this particular case happens with the no-op scheduler, it seems this can happen with the pointwise scheduler as well. Slices are only supported with segment inputs, so prepending unary ops after a segment is accepted seems to break the slice assumption.

Modifying a fusion segment after it gets accepted by a scheduler seems to be the root cause of these errors. This is definitely something to reconsider for future segementer improvement, but for now a work-around would be to consider those excluded unary ops when doing the alias analysis.

CC: @wujingyue

**Comment by wujingyue** [comment]
> A segment with just the slice op is detected as a no-op segment

I'm confused -- didn't the log say that slice-only segment was accepted as pointwise? 

```
**Segmenter** Considering fusion:
i13 = fmin(i0, 5);
i15 = fmax(0, i13);
T2_g[ iS4{( i0 + ( ( fmax(0, ( fmin(i0, 5) )) ) + ( -i0 ) ) )}rf ]
   = slice( T1_g[ iS1{i0} ], { {0, i15, 1} } )

Scheduler _no_op_ ***rejected*** because : output has a concrete dimension
Scheduler _reduction_ ***rejected*** because : No reduction op to schedule
Scheduler _transpose_ ***rejected*** because : cannot find two mismatching inner most dimensions
Scheduler _transpose_ ***rejected*** because : cannot find two mismatching inner most dimensions
***Accepted*** as: pointwise
```

**Comment by wujingyue** [comment]
> Slices are only supported with segment inputs

Speaking of workarounds, can we add a pre-segmentation pass that moves `slice` as early as possible? This seems useful in general because it reduces the tensor size earlier. It also should work around this particular bug because the slice would be moved to the beginning.

**Comment by naoyam** [comment]
> > A segment with just the slice op is detected as a no-op segment
> 
> I'm confused -- didn't the log say that slice-only segment was accepted as pointwise?
> 
> ```
> **Segmenter** Considering fusion:
> i13 = fmin(i0, 5);
> i15 = fmax(0, i13);
> T2_g[ iS4{( i0 + ( ( fmax(0, ( fmin(i0, 5) )) ) + ( -i0 ) ) )}rf ]
>    = slice( T1_g[ iS1{i0} ], { {0, i15, 1} } )
> 
> Scheduler _no_op_ ***rejected*** because : output has a concrete dimension
> Scheduler _reduction_ ***rejected*** because : No reduction op to schedule
> Scheduler _transpose_ ***rejected*** because : cannot find two mismatching inner most dimensions
> Scheduler _transpose_ ***rejected*** because : cannot find two mismatching inner most dimensions
> ***Accepted*** as: pointwise
> ```

Hmm, I was looking at an output when I ran the repro myself with the current main branch, and here's what I got:

```
...
**Segmenter** Considering fusion:
i13 = fmin(i0, 5);
i15 = fmax(0, i13);
T2_g[ iS4{( i0 + ( ( fmax(0, ( fmin(i0, 5) )) ) + ( -i0 ) ) )}rf ]
   = slice( T1_g[ iS1{i0} ], { {0, i15, 1} } )

***Accepted*** as: no_op
...
```

I suppose there's an update on aliasing support after the commit Jacob used (likely by you!)

**Comment by naoyam** [comment]
But this actually confirms:

> While this particular case happens with the no-op scheduler, it seems this can happen with the pointwise scheduler as well.

**Comment by jacobhinkle** [comment]
> a pre-segmentation pass that moves `slice` as early as possible?

This seems great. One thing to note is that upstreaming slice might change IterTypes of upstream tensors. I'm not sure if that would be an issue but we should just consider "slice to broadcast" when designing and testing such a pass...

A similar transform is moving reshape operations up or downstream when possible.

**Comment by jacobhinkle** [comment]
See #1553 @wujingyue. In that case, `PadOp` also has this problem since it requires inputs to be segment inputs as well. In that case, it's slightly complicated too since moving `PadOp` is not as trivial as it is with `SliceOp`. `slice` commutes with unary ops trivially, but for `pad` we have to conjugate by modifying the pad value by the _inverse_ of the op, which might be tricky to do in a general setting, and if we want to upstream `PadOp` through binary ops I am not sure we can do that at all really. Even if we _can_ upstream a `PadOp`, we have the exact opposite motivation from slice: this leads to potentially _larger_ intermediates instead of smaller as is the case with slice.

So I think we still need a way to prevent prepending unary ops in front of these two special ops, at least as a temporary bandaid.

**Comment by jacobhinkle** [comment]
#1556 fixes this issue if we change the repro to use `DataType::Float` instead of `BFloat16`. When using `BFloat16` we actually get a compile error. I'm going to open a separate issue once I isolate that repro.

**Comment by naoyam** [comment]
#1559 should fix this error. It makes sure the excluded unary ops are indeed valid to merge with the consumer segments.

@jacobhinkle, I see you have a couple of PRs. #1558 seems to be complimentary. We should make sure to remember to remove that constraint once the resize support is extended and the input constraint is removed.

**Comment by jacobhinkle** [comment]
Reopening since I think this repro still fails when the dtype is half precision as in the example. The generated kernel and error:
```c++
__global__ void nvfuser_pointwise_f0_c1_r0_g1(Tensor<__bfloat, 1, 1> T0, Tensor<__bfloat, 1, 1> T1) {
  nvfuser_index_t i0;
  i0 = ((nvfuser_index_t)threadIdx.x) + (128LL * ((nvfuser_index_t)blockIdx.x));
  if ((i0 < T0.logical_size[0LL])) {
    __bfloat T3[1];
    T3[0] = 0;
    T3[0]
       = T0[i0];
    __bfloat T4[1];
    T4[0]
       = -T3[0];
    T1[i0]
       = T4[0];
  }
}
/*
CUDA NVRTC compile error: __tmp_kernel_pointwise_f0_c1_r0_g1.cu(9467): error: no operator "-" matches these operands
            operand types are: - <unnamed>::__bfloat
         = -T3[0];
           ^
*/
```
Looks like we are skipping arithmetic casts.
